0.0.4 fix missing first comment in utf8 files

jmoody opened the pull request:
nice repo. 

i came across some cases where parse method would fail to capture the first comment of certain files.  i narrowed the problem down to utf8 encoded strings files (which i believe is the required format for .strings).  

after some more investigation, i figured out that it was the BOM character at position 0 in the utf8 strings files that was causing this

```
 def whole_comment
      /((^\/\*(.+)\*\/)|(^\/\/(.+)))/.match(content).to_s
 end
```

to return false

i added some more specs and decided to rewrite your `Reader::read` to filter out this BOM character. 

along the way, i added a `.gitignore` file and had to comment out some of your `Rakefile` because it was interacting in a bad way with bundle and building/installing the gem.

jmoody committed:
added gitignore file

jmoody committed:
added additional specs for parsing ASCII files.  commented out some code i did not understand in the rake file

jmoody committed:
refactored create_temp_file to take (required) encoding argument

jmoody committed:
cannot reproduce bug using Tempfile + encoding - will try with actual strings files

jmoody committed:
added some tests around encoding

jmoody committed:
solved the problem with BOM at the beginning of strings files

rylev commented on Rakefile:
```diff
@@ -6,10 +6,11 @@ rescue LoadError
-end
+# this was causing a world of hurt
+#require 'rspec/core/rake_task'
+#RSpec::Core::RakeTask.new do |t|
```
Can you go ahead and just remove what you commented out?

jmoody replied:
hello.

as in, delete from the file all together?

rylev replied:
Yes

rylev commented on spec/ascii.strings:
```diff
@@ -0,0 +1,9 @@
+/* No comment provided by engineer. */
```
This and the UTF8 version should be placed in a "fixtures" folder

jmoody replied:
my ruby foo is pretty low - where would such a directory go?

spec/fixtures?

rylev replied:
Yep, that's the one.

jmoody replied:
moved

rylev commented:
Thanks so much for contributing! I really appreciate it. I'll be making some comments. Also, I quickly checked out your branch and tried to run the tests, but got the following error: 

`require': cannot load such file -- parsed_dot_strings (LoadError)

Are you able to run the tests locally?

rylev commented on .gitignore:
```diff
@@ -0,0 +1,29 @@
+.DS_Store
+__MACOSX
+
+# emacs turds
```
Bit harsh no? ;-)

jmoody replied:
emacs turds? :) 

hard to think of them as anything else.

rylev commented on lib/apfel.rb:
```diff
@@ -7,10 +7,12 @@ module Apfel
   # strings object
   def self.parse(file)
     file = read(file)
+    # confirmed that read does remove the first comment in utf
```
I think both of these comments are unnecessary

rylev commented on spec/apfel_parse_ascii_file_spec.rb:
```diff
@@ -0,0 +1,50 @@
+    context 'when given a ASCII DotStrings file'do
+
+      it 'the file should be ascii' do
+        res = `file  -I ./spec/ascii.strings`
```
I would rename the variable to mime_type

jmoody replied:
mime_type - i don't think so.  mime_types are two-part like text/html

i could get behind `encoding` or `file_info`

rylev replied:
encoding is fine

rylev commented on spec/apfel_parse_ascii_file_spec.rb:
```diff
@@ -0,0 +1,50 @@
+
+      it 'the file should be ascii' do
+        res = `file  -I ./spec/ascii.strings`
+        encoding = res.split(/=/).last.gsub!("\n",'')
```
Do you think it would be more readable to do 

``` ruby
mime_type.should include("charset=ascii")
```

If so you can remove this line completely

jmoody replied:
i didn't know `include` worked like that.  <sigh> 
yes, it is more readable

rylev commented on spec/apfel_parse_ascii_file_spec.rb:
```diff
@@ -0,0 +1,50 @@
+        parsed_file.should be_a(Apfel::ParsedDotStrings)
+      end
+
+      #it 'should have the correct keys' do
```
We still need to check that the correct keys are there. I would uncomment these tests and replace the test string with the keys from the fixture.

jmoody replied:
done

rylev commented on spec/apfel_parse_utf8_file_spec.rb:
```diff
@@ -0,0 +1,53 @@
+        parsed_file.should be_a(Apfel::ParsedDotStrings)
+      end
+
+      #it 'should have the correct keys' do
```
Same as the ascii file. We still need to check the keys and values

jmoody replied:
done

rylev commented on spec/apfel_parse_ascii_file_spec.rb:
```diff
@@ -0,0 +1,50 @@
+require 'spec_helper'
```
I would move these into the apfel_spec file. There's no need to have separate files as they are testing the same functionality just under different contexts

jmoody replied:
done

rylev commented on spec/spec_helper.rb:
```diff
@@ -1,13 +1,17 @@
-def create_temp_file(string)
-  temp_file = Tempfile.new('temp')
+
+# added encoding to see if could could reproduce the missing  first comment in
```
Now that the mystery is solved, I think we should remove these comments.

jmoody replied:
done.

rylev commented on spec/apfel_parse_utf8_file_spec.rb:
```diff
@@ -0,0 +1,53 @@
+          parsed_file.comments['boredom'].should == 'No comment provided by engineer.'
+        end
+      end
+
```
remove space

jmoody replied:
done

jmoody committed:
resolving issues raised by rlevick

* removed comments from rakefile
* removed comments from lib/apfel.rb
* fixed require path of parsed_dot_strings in various spec files
  (causing file not found errors in his environment)
* refactored the file-encoding tests against the ascii and utf8
  strings files to be more readable
* refactored the file-encoding test against the utf8 strings file to
  use unix file instead of File.open because File.open specifies an
  encoding to use (defaults to utf8) so the tests were useless for
  ensuring ut8 encoding
* some cosmetic changes

jmoody committed:
moved the spec/*.strings files to spec/fixtures

jmoody committed:
improved tests done on utf8 and ascii strings files

* now testing that the first character of the utf8.strings is the BOM
* added german translations and translator comments to the ascii and
  utf strings files
* now testing the keys and values of the ascii and utf8 strings files

jmoody committed:
moved all the parse tests to apfel_spec.rb

jmoody commented:
##### quoting @rlevick

Also, I quickly checked out your branch and tried to run the tests, but got the following error:

```
`require': cannot load such file -- parsed_dot_strings (LoadError)
```

Are you able to run the tests locally?

###### reply

yes. i can run the tests locally.   

i do so in the terminal: `bundle exec rspec spec` and from within RubyMine.

i did some grepping and found several specs with:

`require 'parsed_dot_strings'`

i replaced them with:

`require 'apfel/parsed_dot_strings'`

maybe that was the problem?

ruby installed with rbenv

```
$ ruby --version
ruby 1.9.3p392 (2013-02-22 revision 39386) [x86_64-darwin12.3.0]
```

jmoody committed:
now using slice! to remove the BOM character (when it exists), rather
than scanning the whole content (string) for the BOM character

rylev commented:
LGTM
